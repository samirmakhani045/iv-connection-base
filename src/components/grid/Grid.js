import React from 'react';
import PropTypes from 'prop-types';

import { Wrapper } from './Grid.styles';

const Grid = (props) => {
  const { width, columns, gap, gutter, align, xs, sm, md, lg, children, ...remainingProps } = props;
  const propsBreakpoints = { xs, sm, md, lg };
  let breakpoints = {};
  ['xs', 'sm', 'md', 'lg'].forEach((breakpoint) => {
    breakpoints[breakpoint] = {
      align: propsBreakpoints[breakpoint]?.align || align,
      columns: propsBreakpoints[breakpoint]?.columns || columns,
      gap: propsBreakpoints[breakpoint]?.gap || gap,
      gutter: propsBreakpoints[breakpoint]?.gutter || gutter,
      width: propsBreakpoints[breakpoint]?.width || width,
    };
  });

  return (
    <Wrapper
      data-testid="grid"
      align={align}
      columns={columns}
      gap={gap}
      gutter={gutter}
      width={width}
      breakpoints={breakpoints}
      {...remainingProps}
    >
      {children}
    </Wrapper>
  );
};

const BreakpointPropType = PropTypes.shape({
  align: PropTypes.oneOf(['left', 'center', 'right']),
  columns: PropTypes.number,
  gap: PropTypes.string,
  gutter: PropTypes.string,
  width: PropTypes.string,
});

Grid.propTypes = {
  /** Children elements to render */
  children: PropTypes.node,
  /** Max width of the grid, e.g. 100%, 200px, 20rem, etc */
  width: PropTypes.string,
  /** Maximum number of columns per row */
  columns: PropTypes.number,
  /** Gap between columns, e.g, 10%, 20px, 2rem, etc */
  gap: PropTypes.string,
  /** Gutter/margin on each side of grid, e.g, 10%, 20px, 2rem, etc */
  gutter: PropTypes.string,
  /** Force content to snap to one of these positions */
  align: PropTypes.oneOf(['left', 'center', 'right']),
  /** Configuration for breakpoint extraSmall */
  xs: BreakpointPropType,
  /** Configuration for breakpoint small */
  sm: BreakpointPropType,
  /** Configuration for breakpoint medium */
  md: BreakpointPropType,
  /** Configuration for breakpoint large */
  lg: BreakpointPropType,
};

Grid.defaultProps = {
  align: 'left',
  columns: 16,
  gap: '2rem',
  gutter: '0',
  width: '144rem',
};

export default Grid;
