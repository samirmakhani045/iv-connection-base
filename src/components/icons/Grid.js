import React from 'react';
import PropTypes from 'prop-types';

import * as colors from '@studio/theme/colors';

const Grid = (props) => {
  const { fill, ...remainingProps } = props;

  return (
    <svg width="1em" height="1em" viewBox="0 0 24 24" fill={fill} {...remainingProps}>
      <path d="M4 8H8V4H4V8ZM10 20H14V16H10V20ZM4 20H8V16H4V20ZM4 14H8V10H4V14ZM10 14H14V10H10V14ZM16 4V8H20V4H16ZM10 8H14V4H10V8ZM16 14H20V10H16V14ZM16 20H20V16H16V20Z" />
    </svg>
  );
};

Grid.propTypes = {
  /** Which color to apply to the icon */
  fill: PropTypes.string,
};

Grid.defaultProps = {
  fill: colors.textPrimary,
};

export default Grid;
