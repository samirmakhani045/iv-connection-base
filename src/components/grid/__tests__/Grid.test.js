import React from 'react';
import { render, screen } from '@testing-library/react';

import Grid from '../Grid';

const Test = (props) => <Grid {...props} />;

describe('<Grid />', () => {
  it('should render correctly', () => {
    const { container } = render(<Test />);

    expect(screen.queryByTestId('grid')).toBeInTheDocument();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should support changing content alignment, number of columns, gap, gutter and width', () => {
    const { container } = render(
      <Test align="right" columns={8} gap="1rem" gutter="1rem" width="100rem" />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should support configuration on different breakpoints', () => {
    const GRID_CONFIG = {
      xs: {
        columns: 4,
        gap: '1rem',
      },
      md: {
        columns: 12,
        gap: '3rem',
      },
    };
    const { container } = render(<Test {...GRID_CONFIG} />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
