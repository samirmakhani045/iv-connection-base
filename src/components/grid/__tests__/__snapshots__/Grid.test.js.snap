// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Grid /> should render correctly 1`] = `
<div
  class="Gridstyles__Wrapper-qir3g0-0 RwFea"
  data-testid="grid"
  width="144rem"
/>
`;

exports[`<Grid /> should support changing content alignment, number of columns, gap, gutter and width 1`] = `
<div
  class="Gridstyles__Wrapper-qir3g0-0 kWlzTO"
  data-testid="grid"
  width="100rem"
/>
`;

exports[`<Grid /> should support configuration on different breakpoints 1`] = `
<div
  class="Gridstyles__Wrapper-qir3g0-0 dPjpBP"
  data-testid="grid"
  width="144rem"
/>
`;
